lib/PuppeteerSharp.Tests/Page/EvaluateTests.cs

```diff
@@ -0,0 +1,127 @@
+                    frameEvaluation = e.Frame.EvaluateFunctionAsync<int>("() => 6 * 7");
+                };
+
+                await page.GoToAsync(TestConstants.EmptyPage);
+                Assert.Equal(42, await frameEvaluation);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldRejectPromiseWithExeption()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                var exception = await Assert.ThrowsAsync<EvaluationFailedException>(() =>
+                {
+                    return page.EvaluateFunctionAsync<object>("() => not.existing.object.property");
+                });
+
+                Assert.NotNull(exception);
+                Assert.True(exception.Message.Contains("not is not defined"));
+            }
+        }
+
+        [Fact]
+        public async Task SouldReturnComplexObjects()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic obj = new
+                {
+                    foo = "bar!"
+                };
+                dynamic result = await page.EvaluateFunctionAsync("a => a", obj);
+                Assert.Equal("bar!", result.foo.ToString());
+            }
+        }
+
+        [Theory]
+        [InlineData("() => NaN", double.NaN)] //ShouldReturnNaN
+        [InlineData("() => -0", -0)] //ShouldReturnNegative0
+        [InlineData("() => Infinity", double.PositiveInfinity)] //ShouldReturnInfinity
+        [InlineData("() => -Infinity", double.NegativeInfinity)] //ShouldReturnNegativeInfinty
+        public async Task BasicEvaluationTest(string script, object expected)
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic result = await page.EvaluateFunctionAsync(script);
+                Assert.Equal(expected, result);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldAcceptNullAsOneOfMultipleParameters()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                dynamic result = await page.EvaluateFunctionAsync<bool>("(a, b) => Object.is(a, null) && Object.is(b, 'foo')", null, "foo");
+                Assert.True(result);
+            }
+        }
+
+        [Fact]
+        public async Task ShouldProperlyIgnoreUndefinedFields()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                var result = await page.EvaluateFunctionAsync<Dictionary<string, object>>("() => ({a: undefined})");
+                Assert.False(result.ContainsKey("a"));
+            }
+        }
+
+        [Fact]
+        public async Task ShouldProperlySerializeNullFields()
+        {
+            using (var page = await Browser.NewPageAsync())
+            {
+                var result = await page.EvaluateFunctionAsync<Dictionary<string, object>>("() => ({a: null})");
+                Assert.True(result.ContainsKey("a"));
```

Comment:
this is good enough for now, maybe once https://github.com/xunit/xunit/issues/1620 is done we could improve this